SCF.py

```diff
@@ -1,191 +1,218 @@
+
+    Arguments:
+        mol_: the PySCF molecule data structure created from Input
+        h_core_: the one electron hamiltonian matrix
+        er_ints_: the 2e electron repulsion integrals
+        Duv_: the density matrix
+
+    Returns:
+        Fuv: The fock matrix
+
+    """
+
+    Fuv = h_core_.copy()  # Takes care of the Huv part of the fock matrix
+    num_aos = mol_.nao    # Number of atomic orbitals, dimension of the mats
+
+    """
+    Replace with your implementation
+
+    Here you will do the summation of the last two terms in the Fock matrix
+    equation involving the two electron Integrals
+
+    Hint: You can do this with explicit loops over matrix indices, which will
+          have many for loops.
+
+    This can also be done with numpy aggregations, bonus points if you
+    implement this with only two loops.
+
+    For example, the first term can be implemented like the following:
+    (er_ints[mu,nu]*Duv).sum()
+    """
+    
+    for u in range(num_aos):
+        for v in range(num_aos):
+            Fuv[u, v] += (Duv_ * er_ints_[u,v]).sum()
+    
+    for u in range(num_aos):
+        for v in range(num_aos):
+            Fuv[u, v] -= (0.5 * Duv_ * er_ints_[u, : ,v]).sum()
+
+    return Fuv
+
+
+def solve_Roothan_equations(Fuv_, Suv_):
+    """
+    solve_Roothan_equations - Solves the matrix equations to determine
+                              the MO coefficients
+
+    Arguments:
+        Fuv_: The Fock matrix
+        Suv_: The overlap matrix
+
+    Returns:
+        mo_energies: an array that contains eigenvalues of the solution
+        mo_coefficients: a matrix of the eigenvectors of the solution
+
+    """
+
+    """
+    Replace with your implementation
+
+    The Roothan Equations, which are of the form FC=SCe can be solved
+    directly from the proper use of scipy.linalg.eigh since this is a
+    symmetric hermitian matrix. Take a look at the documentation for that
+    function and you can implement this in one line.
+    """
+
+
+    mo_energies, mo_coeffs = sp.linalg.eigh(Fuv_, Suv_)
+    
+    return mo_energies.real, mo_coeffs.real
+
+
+def form_density_matrix(mol_, mo_coeffs_):
+    """
+    form_density_matrix - forms the density matrix from the eigenvectors
+
+    Note: the loops are over the number of electrons / 2, not all of the
+    atomic orbitals
+
+    Arguments:
+        mol_: the PySCF molecule data structure created from Input
+        mo_coeffs_: a matrix of the eigenvectors of the solution
+
+    Returns:
+        Duv: the density matrix
+    """
+
+    nelec = mol_.nelec[0]  # Number of occupied orbitals
+    num_aos = mol_.nao  # Number of atomic orbitals, dimensions of the mats
+    Duv = np.zeros(mol_.nao, mol_.nao, dtype=np.double)
+
+    """
+    Replace with your implementation
+
+    This involves basically a computation of each density matrix element
+    that is a sum over the produces of the mo_coeffs.
+
+    """
+
+    for u in range(num_aos):
```

Comment:
Note, this could be done with only two loops and then a numpy summation.

```python
Duv[u][v] = 2.0*(mo_coeffs_[mu, :nelec]
                               * mo_coeffs_[nu, :nelec]).sum()
 ```